File: rag/advanced_rag/think_log.py
#
#  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

"""Surface selected internal INFO logs to the client as ``<think>`` content.

During an agentic ``rag_agent`` turn we attach a context-scoped logging sink so
the pipeline's bracket-tagged progress logs — ``[Agentic RAG]``,
``[Formalizing the question]``, ``[Preliminary search]``, ``[Planner]``,
``[Orchestrator]``, ``[Agentic research]``, ``[Hybrid search]``,
``[BM25 search]``, ``[Web search]``, ``[Composing the answer]``,
``[Tool loop]``, ``[Function tool]`` … — can be streamed to the front end as
reasoning without instrumenting every call site. The tags double as
human-readable stage labels, so the same message serves the backend log and
the user-facing thinking stream.

The sink is stored in a :class:`contextvars.ContextVar`, so only the async task
tree of the current request (which inherits the context) forwards its logs —
concurrent requests stay isolated. Logs emitted from ``thread_pool_exec``
workers that did not inherit the context are simply skipped.
"""

from __future__ import annotations

import contextvars
import logging
from typing import Callable

# Per-request sink: a callable(str) that forwards one log line, or None when no
# agentic turn is streaming in the current context.
_think_log_sink: contextvars.ContextVar[Callable[[str], None] | None] = contextvars.ContextVar("think_log_sink", default=None)

# Only bracket-tagged INFO lines from these logger namespaces are surfaced.
_SCOPED_PREFIXES = ("rag.advanced_rag", "rag.llm.chat_model", "rag.llm.tool_decorator")

_installed = False


class ThinkLogHandler(logging.Handler):
    """Forward in-scope, bracket-tagged records to the active per-request sink."""

    def emit(self, record: logging.LogRecord) -> None:
        sink = _think_log_sink.get()
        if sink is None:
            return
        name = record.name or ""
        if not name.startswith(_SCOPED_PREFIXES):
            return
        try:
            msg = record.getMessage()
        except Exception:
            return
        # Only the bracket-tagged progress lines ("[Hybrid search] ...").
        if not msg or not msg.lstrip().startswith("["):
            return
        try:
            sink("<br>" + msg.strip())
        except Exception:
            # Never let think-log forwarding break the request or the logging
            # subsystem itself.
            pass


def install_think_log_handler() -> None:
    """Install the forwarding handler on the root logger exactly once."""
    global _installed
    if _installed:
        return
    handler = ThinkLogHandler(level=logging.INFO)
    logging.getLogger().addHandler(handler)
    _installed = True


def set_think_log_sink(sink: Callable[[str], None] | None):
    """Activate ``sink`` for the current context; returns the reset token."""
    return _think_log_sink.set(sink)


def reset_think_log_sink(token) -> None:
    try:
        _think_log_sink.reset(token)
    except Exception:
        pass

File: internal/rag/advanced_rag/think_log.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package advanced_rag

import (
	"io"
	"log"
	"strings"
)

// Surface selected internal stage logs to the client as thinking content.
//
// Bracket-tagged lines such as "[Agentic RAG]", "[Planner]", "[Orchestrator]",
// "[Hybrid search]" and "[Composing the answer]" are forwarded to a per-request sink, so
// the front end shows live reasoning WITHOUT instrumenting every call site. The tags
// double as human-readable stage labels, so one message serves both the backend log and
// the thinking stream.
//
// The interception happens at the WRITER the run's logger is rebuilt on top of, because
// the log package has no per-record handler hook. That logger is threaded into SearchDeps
// and every graph node, so wrapping it once in Rag covers the whole pipeline.
//
// Two consequences of filtering on the tag SHAPE alone:
//   - numbered evidence markers ("[1] chunk text") have to be rejected: they are chunk
//     listings, not stage tags (see isEvidenceMarker);
//   - the tool narration ("[Function tool] Running the {name} tool with: …") reaches the
//     block by itself, because harness/tool_executor.go logs through the same wrapped
//     logger. The line around a single harness invocation comes from
//     internal/service/chat_pipeline.go (toolLoopLine), which has no model-authored loop
//     to instrument.
//
// The think block is HTML, so a stage line ends with ThinkLineBreak rather than "\n" (see
// ThinkLineBreak).

// ThinkLineBreak separates two lines inside the think block.
//
// The block is delivered to the client as inline HTML (the chat UI wraps it in
// <details class="think"> and renders the message as markdown), where a newline is just
// whitespace — so every think-line producer (thinkWriter, the outer react loop's progress
// routing, chat_pipeline's tool-loop narration) must use this marker, or consecutive
// stage lines merge into one sentence.
const ThinkLineBreak = "<br>"

// thinkWriter mirrors one log line onto the real logger while forwarding
// bracket-tagged stage lines to the per-request progress sink.
type thinkWriter struct {
	orig *log.Logger // the run's real logger; keeps backend logging unchanged
	sink func(string)
}

// Write implements io.Writer.
func (w thinkWriter) Write(p []byte) (int, error) {
	line := strings.TrimRight(string(p), "\n")
	if line == "" {
		return len(p), nil
	}
	// The filter is the tag SHAPE: a leading "[" that is not a numbered evidence marker
	// such as "[1] ...", which is a chunk listing rather than a stage tag.
	if trimmed := strings.TrimSpace(line); strings.HasPrefix(trimmed, "[") && !isEvidenceMarker(trimmed) {
		w.forward(trimmed)
	}
	// Re-emit through the ORIGINAL logger so its own routing (zap) and level are
	// preserved. This writer is built with no prefix and no flags, so the
	// backend output is exactly what it was before the wrap.
	if w.orig != nil {
		w.orig.Printf("%s", line)
	}
	return len(p), nil
}

// forward delivers one line to the sink.
//
// The separator is ThinkLineBreak: the thinking block reaches the client as the inline
// "<think>...<details class="think">" HTML that the markdown renderer passes through,
// where a bare "\n" collapses into a space — every stage line would then run together
// into one unreadable sentence.
//
// The recover() guards the sink: forwarding a think line must never break the request or
// the logging subsystem itself.
func (w thinkWriter) forward(line string) {
	defer func() { _ = recover() }()
	if w.sink == nil {
		return
	}
	// The break is TRAILING here: prefixing it would open the block with a blank line.
	w.sink(line + ThinkLineBreak)
}

// isEvidenceMarker reports whether line is a numbered chunk/option listing such
// as "[1] passage text" rather than a stage tag like "[Planner] ...".
func isEvidenceMarker(line string) bool {
	rest := line[1:]
	if rest == "" {
		return false
	}
	n := 0
	for n < len(rest) && rest[n] >= '0' && rest[n] <= '9' {
		n++
	}
	return n > 0 && n < len(rest) && rest[n] == ']'
}

// thinkLogger returns a *log.Logger that logs through orig while additionally
// forwarding bracket-tagged stage lines to sink. The sink is per REQUEST, so concurrent
// requests stay isolated.
//
// sink may be nil, in which case orig is returned unchanged.
func thinkLogger(orig *log.Logger, sink func(string)) *log.Logger {
	if orig == nil || sink == nil {
		return orig
	}
	return log.New(thinkWriter{orig: orig, sink: sink}, "", 0)
}

var _ io.Writer = thinkWriter{}
